File: java/src/main/java/io/github/messageflow/CancellationToken.java
package io.github.messageflow;

import java.util.concurrent.CancellationException;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Propagates a cancellation request through a chain.
 *
 * <p>Tokens are created by a {@link CancellationTokenSource}; {@link #none()} returns a token that is
 * never cancelled.
 */
public final class CancellationToken {

    private static final CancellationToken NONE = new CancellationToken(null);

    private final AtomicBoolean cancelled;

    CancellationToken(AtomicBoolean cancelled) {
        this.cancelled = cancelled;
    }

    /**
     * Returns a token that is never cancelled.
     *
     * @return the non-cancellable token
     */
    public static CancellationToken none() {
        return NONE;
    }

    /**
     * Indicates whether cancellation has been requested.
     *
     * @return {@code true} when the source of this token was cancelled
     */
    public boolean isCancellationRequested() {
        return cancelled != null && cancelled.get();
    }

    /**
     * Throws when cancellation has been requested.
     *
     * @throws CancellationException cancellation has been requested
     */
    public void throwIfCancellationRequested() {
        if (isCancellationRequested()) {
            throw new CancellationException("The operation was cancelled.");
        }
    }
}

File: node/src/CancellationToken.ts
/**
 * Propagates a cancellation request through a chain.
 *
 * Tokens are created by a {@link CancellationTokenSource}; {@link CancellationToken.none}
 * returns a token that is never cancelled.
 */
export class CancellationToken {
  private static readonly NONE = new CancellationToken(null);

  private constructor(private readonly cancelled: { value: boolean } | null) {}

  /**
   * Returns a token that is never cancelled.
   *
   * @returns the non-cancellable token
   */
  static none(): CancellationToken {
    return CancellationToken.NONE;
  }

  /**
   * Indicates whether cancellation has been requested.
   *
   * @returns `true` when the source of this token was cancelled
   */
  isCancellationRequested(): boolean {
    return this.cancelled !== null && this.cancelled.value;
  }

  /**
   * Throws when cancellation has been requested.
   *
   * @throws {Error} with name 'CancellationError' when cancellation has been requested
   */
  throwIfCancellationRequested(): void {
    if (this.isCancellationRequested()) {
      const error = new Error('The operation was cancelled.');
      error.name = 'CancellationError';
      throw error;
    }
  }

  /** @internal */
  static createFromSource(cancelled: { value: boolean }): CancellationToken {
    return new CancellationToken(cancelled);
  }
}
